0000-experiment-dataframe.md

```diff
@@ -0,0 +1,320 @@
+- chisq (float): Same as the proposed `chisq` field.
+- device_components (List[DeviceComponent]): Same as the proposed `device_components` field.
+- quality (str): One of three possible values: `Bad`, `No Information`, `Good`. This is slightly different than the strings being used in QE.
+- tags (List[str]): Same as the proposed `tags` field.
+- verified (bool): Same as the proposed `verified` field.
+- device_name (str): Same as the proposed `backend` field.
+- created_at (string): An ISO 8601 timestamp indicating the creation time of the analysis
+result inside the resultDB. This is similar to the `creation_datetime` field but not the same.
+Usually experimentalists are interested in the time at which the experiment was done, rather than a timestamp for the resultDB entry.
+This suggests these time information must be handled separately.
+- fit (dict): A free-form dictionary allowing the storage of arbitrary data.
+This can be used to store the `value`, `extra` and `source` fields.
+
+Two of the proposed fields do not make sense in a resultDB context and won't be saved/loaded but only handled by QE:
+- `created_in_db`: This has been conventionally managed internally in the `ExperimentData`.
+- `experiment`: This may be stored in the `fit` field. Not only experiment UUID, but also human readable representation of experiment is important to help experimentalists with filtering and sorting analysis parameters.
+
+To reduce the amount of work done by QE, we suggest that the experiment service `qiskit-ibm-experiment` will implement
+the capability to work with pandas data frames in the suggested structure, and perform the following changes:
+- Breakdown the data frame to individual results.
+- Translate the fields
+   - `uuid` <--> `result_id`
+   - `experiment_uuid` <--> `experiment_id`
+   - `type` <--> `name`
+   - `device_name` <--> `backend`
+   - `quality` (db format) <--> `quality` (QE format)
+   - `fit` <--> `value`, `extra`, `source`
+   - TODO: still need to determine the use of `creation_datetime`
+- Efficiently upload/download multiple results from the server. This depends on the
+API supplied by the server; currently mass loading/updating is possible in one API call,
+but mass creation is not supported, meaning the service might have to resort to multithreading.
+
+To preserve backward compatibility, we suggest all data frame handling in `qiskit-ibm-experiment` to be
+indicted using an optional parameter, `dataframe` (bool) which defaults to `False`, inside the methods
+
+- `IBMExperimentService.create_analysis_results()`
+- `IBMExperimentService.update_analysis_results()`
+- `IBMExperimentService.analysis_result()`
+- `IBMExperimentService.analysis_results()`
+
+## Migration plan
+
+Even though this is the change of internal data structure, this may impact community developers who write own experiment on top of our base classes. Thus, the change should be made step by step to guarantee the backward compatibility.
```

Comment:
I think here is where we should try and have fixed classes for the service/db for relevant data types, and the intention is these should not be modified/subclasses, but worked with by other classes within qiskit-experiments or other frameworks.

Reply:
^ @gadial

Reply:
Currently `qiskit-ibm-experiment` uses the `ExperimentData` and `AnalysisResultData` dataclasses, which are also used inside `qiskit-experiments`. I have no objection to modifying how these dataclasses look or used, but this is really a question of what `qiskit-experiments` users would use.

We can also drop these classes altogether and stick with padas dataframes (that's basically what the local DB is doing now). Again, the question is what users would prefer.